File: BinaryTree/python/leafList.py
from typing import List, Optional

from node import Node


def leaf_list(root: Optional[Node]) -> List[int]:
    leaves: List[int] = []
    collect(root, leaves)
    return leaves


def collect(node: Optional[Node], leaves: List[int]) -> None:
    if node is None:
        return
    if node.left is None and node.right is None:
        leaves.append(node.val)
        return
    collect(node.left, leaves)
    collect(node.right, leaves)

File: BinaryTree/csharp/LeafList.cs
using System.Collections.Generic;

namespace BinaryTreeSolutions;

public static class LeafList
{
    public static IList<int> Solve(TreeNode? root)
    {
        var leaves = new List<int>();
        Collect(root, leaves);
        return leaves;
    }

    private static void Collect(TreeNode? node, ICollection<int> leaves)
    {
        if (node is null)
        {
            return;
        }

        if (node.Left is null && node.Right is null)
        {
            leaves.Add(node.Val);
            return;
        }

        Collect(node.Left, leaves);
        Collect(node.Right, leaves);
    }
}
